src/app/model-poller.js

```diff
@@ -21,46 +21,72 @@ import {
 
 import { userIsControllerAdmin } from "./utils";
 
-export default async function connectAndListModels(reduxStore, bakery) {
+export default async function connectAndListModels(
+  reduxStore,
+  bakery,
+  additionalControllers
+) {
   try {
     const storeState = reduxStore.getState();
-    const credentials = getUserPass(storeState);
     const { identityProviderAvailable, isJuju } = getConfig(storeState);
     const wsControllerURL = getWSControllerURL(storeState);
-    const { conn, error, juju, intervalId } = await loginWithBakery(
+    const credentials = getUserPass(wsControllerURL, storeState);
+    const defaultControllerData = [
       wsControllerURL,
       credentials,
       bakery,
-      identityProviderAvailable
-    );
-    if (error) {
-      reduxStore.dispatch(storeLoginError(error));
-      return;
+      identityProviderAvailable,
+    ];
+    let controllerList = [defaultControllerData];
+    if (additionalControllers) {
+      controllerList = controllerList.concat(additionalControllers);
     }
-    if (process.env.NODE_ENV === "production") {
-      Sentry.setTag("jujuVersion", conn?.info?.serverVersion);
-    }
-    reduxStore.dispatch(updateControllerConnection(conn));
-    reduxStore.dispatch(updateJujuAPIInstance(juju));
-    reduxStore.dispatch(updatePingerIntervalId(intervalId));
-    if (userIsControllerAdmin(conn)) {
-      fetchControllerList(conn, reduxStore);
-      if (!isJuju) {
-        // This call will be a noop if the user isn't an administrator
-        // on the JIMM controller we're connected to.
-        disableControllerUUIDMasking(conn);
+    controllerList.forEach(async (controllerData) => {
+      const { conn, error, juju, intervalId } = await loginWithBakery(
+        ...controllerData
+      );
+
+      if (error) {
+        reduxStore.dispatch(storeLoginError(error));
+        return;
       }
-    }
-    do {
-      await reduxStore.dispatch(fetchModelList());
-      await fetchAllModelStatuses(conn, reduxStore);
-      // Wait 30s then start again.
-      await new Promise((resolve) => {
-        setTimeout(() => {
-          resolve(true);
-        }, 30000);
-      });
-    } while (isLoggedIn(reduxStore.getState()));
+
+      if (process.env.NODE_ENV === "production") {
+        Sentry.setTag("jujuVersion", conn?.info?.serverVersion);
+      }
+
+      reduxStore.dispatch(updateControllerConnection(controllerData[0], conn));
+      reduxStore.dispatch(updateJujuAPIInstance(controllerData[0], juju));
+      reduxStore.dispatch(
+        updatePingerIntervalId(controllerData[0], intervalId)
+      );
+      if (userIsControllerAdmin(conn)) {
+        fetchControllerList(
+          controllerData[0],
+          conn,
+          controllerData[4],
+          reduxStore
+        );
+        // XXX the isJuju Check needs to be done on a per-controller basis
+        if (!isJuju) {
+          // This call will be a noop if the user isn't an administrator
```

Comment:
a noop?

Reply:
It means "no operation". Something is a noop if it doesn't do anything.